src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -0,0 +1,241 @@
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.RegularAndStaticColumns;
+import org.apache.cassandra.db.WriteContext;
+import org.apache.cassandra.db.filter.RowFilter;
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.db.lifecycle.Tracker;
+import org.apache.cassandra.db.memtable.Memtable;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.sai.metrics.TableQueryMetrics;
+import org.apache.cassandra.index.sai.metrics.TableStateMetrics;
+import org.apache.cassandra.index.sai.plan.StorageAttachedIndexQueryPlan;
+import org.apache.cassandra.index.transactions.IndexTransaction;
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SSTableFlushObserver;
+import org.apache.cassandra.notifications.INotification;
+import org.apache.cassandra.notifications.INotificationConsumer;
+import org.apache.cassandra.notifications.MemtableDiscardedNotification;
+import org.apache.cassandra.notifications.MemtableRenewedNotification;
+import org.apache.cassandra.schema.TableMetadata;
+
+/**
+ * Orchestrates building of storage-attached indices, and manages lifecycle of resources shared between them.
+ */
+@ThreadSafe
+public class StorageAttachedIndexGroup implements Index.Group, INotificationConsumer
+{
+    private final TableQueryMetrics queryMetrics;
+    private final TableStateMetrics stateMetrics;
+    private final Set<StorageAttachedIndex> indices = Sets.newConcurrentHashSet();
+    private final ColumnFamilyStore baseCfs;
+
+    StorageAttachedIndexGroup(ColumnFamilyStore baseCfs)
+    {
+        this.baseCfs = baseCfs;
+        this.queryMetrics = new TableQueryMetrics(baseCfs.metadata());
+        this.stateMetrics = new TableStateMetrics(baseCfs.metadata(), this);
+
+        Tracker tracker = baseCfs.getTracker();
+        tracker.subscribe(this);
+    }
+
+    @Nullable
+    public static StorageAttachedIndexGroup getIndexGroup(ColumnFamilyStore cfs)
+    {
+        return (StorageAttachedIndexGroup) cfs.indexManager.getIndexGroup(StorageAttachedIndexGroup.class);
+    }
+
+    @Override
+    public Set<Index> getIndexes()
+    {
+        return ImmutableSet.copyOf(indices);
+    }
+
+    @Override
+    public void addIndex(Index index)
+    {
+        assert index instanceof StorageAttachedIndex;
+        indices.add((StorageAttachedIndex) index);
+    }
+
+    @Override
+    public void removeIndex(Index index)
+    {
+        assert index instanceof StorageAttachedIndex;
+        boolean removed = indices.remove(index);
+        assert removed : "Cannot remove non-existing index " + index;
+        /*
+         * For the in-memory implementation we only need to unsubscribe from the tracker
+         */
+        if (indices.isEmpty())
+        {
+            baseCfs.getTracker().unsubscribe(this);
```

Comment:
Should we just call `invalidate()` here so we release the metrics? It seems like once we get here, there should be no way we ever add anything to the index group again (given SIM immediately removes the group...although we don't directly enforce this).

Reply:
I'd prefer to leave this till later because this changed quite a lot when we add on-disk.